internal/agentcfg/kibana.go

```diff
@@ -0,0 +1,98 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agentcfg
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/apm-server/internal/kibana"
+)
+
+// Error Messages used to signal fetching errors
+const (
+	ErrMsgReadKibanaResponse = "unable to read Kibana response body"
+	ErrMsgSendToKibanaFailed = "sending request to kibana failed"
+	ErrUnauthorized          = "Unauthorized"
+)
+
+const endpoint = "/api/apm/settings/agent-configuration/search"
+
+// KibanaFetcher holds static information and information shared between requests.
+// It implements the Fetch method to retrieve agent configuration information.
+type KibanaFetcher struct {
+	*cache
+	logger *logp.Logger
+	client *kibana.Client
+}
+
+// NewKibanaFetcher returns a KibanaFetcher instance.
+//
+// NewKibanaFetcher will panic if passed a nil client.
+func NewKibanaFetcher(client *kibana.Client, cacheExpiration time.Duration) *KibanaFetcher {
+	if client == nil {
+		panic("client is required")
```

Comment:
**issue**: panic outside test code feels a bit dangerous. Can we return an error instead ?

Reply:
Same, it may not be ideal but this is existing code: https://github.com/elastic/apm-server/blob/main/internal/agentcfg/fetch.go#L68

Reply:
Added #9929